File: 3548.equal-sum-grid-partition-ii.cpp
#
# @lc app=leetcode id=3548 lang=cpp
#
# [3548] Equal Sum Grid Partition II
#

# @lc code=start
class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        const int MAXV = 100000;
        vector<long long> prefix_row(m + 1, 0);
        for (int i = 0; i < m; ++i) {
            long long rowsum = 0;
            for (int j = 0; j < n; ++j) {
                rowsum += grid[i][j];
            }
            prefix_row[i + 1] = prefix_row[i] + rowsum;
        }
        long long S = prefix_row[m];
        vector<long long> prefix_col(n + 1, 0);
        for (int j = 0; j < n; ++j) {
            long long colsum = 0;
            for (int i = 0; i < m; ++i) {
                colsum += grid[i][j];
            }
            prefix_col[j + 1] = prefix_col[j] + colsum;
        }
        vector<vector<int>> row_pos(MAXV + 1);
        vector<vector<int>> col_pos(MAXV + 1);
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int v = grid[i][j];
                row_pos[v].push_back(i);
                col_pos[v].push_back(j);
            }
        }
        for (int v = 0; v <= MAXV; ++v) {
            sort(row_pos[v].begin(), row_pos[v].end());
            sort(col_pos[v].begin(), col_pos[v].end());
        }
        // Horizontal cuts
        for (int k = 1; k < m; ++k) {
            long long sum_top = prefix_row[k];
            long long sum_bot = S - sum_top;
            if (sum_top == sum_bot) return true;
            if (sum_top > sum_bot) {
                long long diff = sum_top - sum_bot;
                if (diff > MAXV) continue;
                int h = k, w = n;
                bool thick = (h >= 2 && w >= 2);
                int vv = (int)diff;
                if (thick) {
                    auto& lis = row_pos[vv];
                    auto it = lower_bound(lis.begin(), lis.end(), k);
                    if (it != lis.begin()) return true;
                } else {
                    if (w == 1) {
                        if ((long long)grid[0][0] == diff || (long long)grid[k - 1][0] == diff) return true;
                    } else {
                        if ((long long)grid[0][0] == diff || (long long)grid[0][n - 1] == diff) return true;
                    }
                }
            } else if (sum_bot > sum_top) {
                long long diff = sum_bot - sum_top;
                if (diff > MAXV) continue;
                int h = m - k, w = n;
                bool thick = (h >= 2 && w >= 2);
                int vv = (int)diff;
                if (thick) {
                    auto& lis = row_pos[vv];
                    auto it = lower_bound(lis.begin(), lis.end(), k);
                    if (it != lis.end()) return true;
                } else {
                    if (w == 1) {
                        if ((long long)grid[k][0] == diff || (long long)grid[m - 1][0] == diff) return true;
                    } else {
                        if ((long long)grid[m - 1][0] == diff || (long long)grid[m - 1][n - 1] == diff) return true;
                    }
                }
            }
        }
        // Vertical cuts
        for (int k = 1; k < n; ++k) {
            long long sum_left = prefix_col[k];
            long long sum_right = S - sum_left;
            if (sum_left == sum_right) return true;
            if (sum_left > sum_right) {
                long long diff = sum_left - sum_right;
                if (diff > MAXV) continue;
                int hh = m, ww = k;
                bool thick = (hh >= 2 && ww >= 2);
                int vv = (int)diff;
                if (thick) {
                    auto& lis = col_pos[vv];
                    auto it = lower_bound(lis.begin(), lis.end(), k);
                    if (it != lis.begin()) return true;
                } else {
                    if (hh == 1) {
                        if ((long long)grid[0][0] == diff || (long long)grid[0][k - 1] == diff) return true;
                    } else {
                        if ((long long)grid[0][0] == diff || (long long)grid[m - 1][0] == diff) return true;
                    }
                }
            } else if (sum_right > sum_left) {
                long long diff = sum_right - sum_left;
                if (diff > MAXV) continue;
                int hh = m, ww = n - k;
                bool thick = (hh >= 2 && ww >= 2);
                int vv = (int)diff;
                if (thick) {
                    auto& lis = col_pos[vv];
                    auto it = lower_bound(lis.begin(), lis.end(), k);
                    if (it != lis.end()) return true;
                } else {
                    if (hh == 1) {
                        if ((long long)grid[0][k] == diff || (long long)grid[0][n - 1] == diff) return true;
                    } else {
                        if ((long long)grid[0][n - 1] == diff || (long long)grid[m - 1][n - 1] == diff) return true;
                    }
                }
            }
        }
        return false;
    }
};
# @lc code=end

File: 3548.equal-sum-grid-partition-ii.java
#
# @lc app=leetcode id=3548 lang=java
#
# [3548] Equal Sum Grid Partition II
#
# @lc code=start
class Solution {
    public boolean canPartitionGrid(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        long[] rowSum = new long[m];
        long[] colSum = new long[n];
        long total = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                rowSum[i] += grid[i][j];
                colSum[j] += grid[i][j];
                total += grid[i][j];
            }
        }
        // Horizontal cuts
        long sumTop = 0;
        for (int cut = 0; cut < m - 1; ++cut) {
            sumTop += rowSum[cut];
            long sumBot = total - sumTop;
            if (sumTop == sumBot) return true;
            long diff = Math.abs(sumTop - sumBot);
            // Consider discounting a border cell and check constraints
            for (int j = 0; j < n; ++j) {
                // Discount cell from top section's border
                if (sumTop > sumBot && grid[cut][j] == diff && isSectionConnectedAfterRemoval(grid, 0, 0, cut, n, cut, j)) return true;
                // Discount cell from bottom section's border
                if (sumBot > sumTop && grid[cut+1][j] == diff && isSectionConnectedAfterRemoval(grid, cut+1, 0, m, n, cut+1, j)) return true;
            }
        }
        // Vertical cuts
        long sumLeft = 0;
        for (int cut = 0; cut < n - 1; ++cut) {
            sumLeft += colSum[cut];
            long sumRight = total - sumLeft;
            if (sumLeft == sumRight) return true;
            long diff = Math.abs(sumLeft - sumRight);
            // Consider discounting a border cell and check constraints
            for (int i = 0; i < m; ++i) {
                // Discount cell from left section's border
                if (sumLeft > sumRight && grid[i][cut] == diff && isSectionConnectedAfterRemoval(grid, 0, 0, m, cut+1, i, cut)) return true;
                // Discount cell from right section's border
                if (sumRight > sumLeft && grid[i][cut+1] == diff && isSectionConnectedAfterRemoval(grid, 0, cut+1, m, n, i, cut+1)) return true;
            }
        }
        return false;
    }
    // General helper to check if removing cell (remRow, remCol) from section [row0,row1)x[col0,col1) leaves it connected
    private boolean isSectionConnectedAfterRemoval(int[][] grid, int row0, int col0, int row1, int col1, int remRow, int remCol) {
        int m = row1 - row0, n = col1 - col0;
        boolean[][] visited = new boolean[m][n];
        int cnt = 0, total = 0;
        int[] dr = {-1,1,0,0};
        int[] dc = {0,0,-1,1};
        // Count how many cells in section after removal
        for (int i = row0; i < row1; ++i) for (int j = col0; j < col1; ++j) if (!(i == remRow && j == remCol)) ++total;
        // Start BFS from any cell in section besides the removed cell
        outer: for (int i = row0; i < row1; ++i)
            for (int j = col0; j < col1; ++j)
                if (!(i == remRow && j == remCol)) {
                    java.util.Queue<int[]> q = new java.util.LinkedList<>();
                    q.offer(new int[]{i-row0, j-col0});
                    visited[i-row0][j-col0] = true;
                    cnt = 1;
                    while (!q.isEmpty()) {
                        int[] cur = q.poll();
                        for (int d = 0; d < 4; ++d) {
                            int ni = cur[0]+dr[d], nj = cur[1]+dc[d];
                            int ai = ni+row0, aj = nj+col0;
                            if (0 <= ni && ni < m && 0 <= nj && nj < n && !(ai == remRow && aj == remCol) && !visited[ni][nj]) {
                                visited[ni][nj] = true;
                                q.offer(new int[]{ni, nj});
                                ++cnt;
                            }
                        }
                    }
                    break outer;
                }
        return cnt == total;
    }
}
# @lc code=end